### PR TITLE
fix(trade): take time offset into account for quote and order validTo

### LINKED ISSUE #4030: Error loading price issue when time is incorrect on a device
Reported in https://cowservices.slack.com/archives/C03NXGFJA2U/p1710175871205829
Discussed in https://cowservices.slack.com/archives/C0361CDG8GP/p1710232016500759

**To reproduce:** 
1. change time on your device (I set +3 hours)
2. Open CoW Swap and try placing a trade

**AR**: impossible: error loading price error is displayed

**ER**: the quote should not rely on the local PC time.

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #4030

Since users local time might be different to the real time, we have to take in into account while `order.validTo` calculating and quote expiration checking.
First of all let's look at typical quote API request and response:

Request

```
{
    ...
    validFor: 2700
}
```

Response

```
{
    expiration: "2024-04-15T08:04:54.698863496Z"
    ...
    quote: {
       validTo: 1713170874
    }
}
```

### ValidFor
In the request we specify `validFor` (duration in seconds) which means time to life of an order.
For quotes in Swap we take the value from Swap settings ("Swap deadline").
For quotes in Limit and TWAP we use hardcoded value `DEFAULT_QUOTE_TTL` (30 min) because for those trading type orders don't have to contain the same `validTo` value as in quote.
Also, when `validFor` and `validTo` are not specified in the quote request, the backend will use `30 mins` as default.

### Expiration
This is actually a quote expiration time. The value is an UTC date in ISO string format.
Usually, the date is 2 minutes after the quote request time, it means that the quote time to life ~2 minutes.

### Quote.validTo
In the quote response we have `validTo` value which is a timestamp of a supposed order. The value equals `Date.now() + validFor`, but what is important `Date.now()` is taken on the backend side which is supposed to be real time. This value helps us to calculate the time difference between user's local time and backend time (real time).

### What is done

1. As it turned out, quote API supports `validFor` parameter, [so I replaced `validTo` parameter by it](https://github.com/cowprotocol/cowswap/pull/4236/commits/7b374575f1effc67ff950149e3c8fe3e8e60a463). Because when local time is not sync with the backend time we can get "validTo is too far into the future" error from the API.

2. For time difference calculation we need three values: when quote was created (`quoteDate`), order ttl (`validFor`), supposed order ttl (`quote.validTo`) which is calculated on backend side. Because for it, I've added:
 - `quoteParams` to the `TradeQuoteState` (for Limit and TWAP) ([commit](https://github.com/cowprotocol/cowswap/pull/4236/commits/8d8bffb79044842174434df32a9af32cd1587cc4))
 - `quoteValidTo` to `QuoteInformationObject` (only for Swap) ([commit](https://github.com/cowprotocol/cowswap/pull/4236/commits/3e1e1a995344a77ab675025b5d98f4b3dfcab47a))

3. We have a state in trade widgets for the case when quote is expired. The state calculates basing on the local device time, so we have to take time offset into account. Because of it, I added the time offset into `isQuoteExpired()` ([commit](https://github.com/cowprotocol/cowswap/pull/4236/commits/805fabc9df6635c0756d7acb308872162f44f24e))

4. And final change - `validTo` in the order data we sign. The value used to be calculated in `calculateValidTo()` function which doesn't take time offset into account. I've replaced the function by `getOrderValidTo()` which includes time offset. ([commit](https://github.com/cowprotocol/cowswap/pull/4236/commits/4b20a665ae4a22ff341b119499c2389464e2e45c))

# To Test

1. Move your local device time in the future (+3 hours for example)
2. Create a swap
- [ ] quote request should contain `validFor` value
- [ ] quote response should contain correct `validTo` value
3. Sign and post an order
4. Open the order in Explorer
- [ ] Expiration time should correspond to values you entered (deadline in swap settings)
5. The same checks for Limit and Twap
6. Eth-flow
7. Limit order in Safe with approve tx bundling

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteDeadline.test.ts
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteDeadline.test.ts:1]: Can you add unit tests for the other fns on quoteDeadline?

**Comment by elena-zh** [comment]
Hey @shoom3301 , I have placed a swap order, but the UI says that it is [expired](https://dev.explorer.cow.fi/sepolia/orders/0x66d7a1c1074ad729efbad8fe19d6628874bf9455815245f9b2b41155dab913d9ff714b8b0e2700303ec912bd40496c3997ceea2b661d0f0e?tab=overview): 
activity modal
placed
image

Details: I moved time 4 hours ahead, swap order duration is set 5 mins.

**Comment by elena-zh** [comment]
For a limit [order](https://dev.explorer.cow.fi/sepolia/orders/0x520a35b0878f57e66b3827b3c72546df0db88039d4c90e30e601476e78ded51cff714b8b0e2700303ec912bd40496c3997ceea2b661d1800?tab=overview): I can place it, however, I see both Placed+Expired notifications, and I see 'ValidFor' in the quote request 
image
image
Duration is correct, timing corresponds to the current timezone (but not +4 hors)

**Comment by elena-zh** [comment]
ETH order is also traded as expired: 
image
image
image
image

**Comment by elena-zh** [comment]
TWAP: is successfully created, but then is moved to Expired state into History section. There is changes statuses to Filled (and other states)
image
image

**Comment by shoom3301** [comment]
image

@elena-zh as we discussed, added the banner when local device time is not valid.
The banner is displayed only when time gap between local and real more than 60 seconds.
You can see the time offset in the console log:

image

File: apps/cowswap-frontend/src/legacy/state/price/reducer.ts [hunk 1]
```diff
@@ -139,7 +141,11 @@ export default createReducer(initialState, (builder) =>
       const shouldUpdate = !(!isBestQuote && hasPrice)
 
       if (quoteInformation && shouldUpdate) {
-        quotes[chainId][sellToken] = { ...quoteInformation, ...payload }
+        quotes[chainId][sellToken] = {
+          ...quoteInformation,
+          ...payload,
+          quoteDate: Math.ceil(Date.now() / 1000),
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply apps/cowswap-frontend/src/legacy/state/price/reducer.ts]: Renamed

File: apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteDeadline.test.ts
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteDeadline.test.ts:1]: Added

**Review by elena-zh — APPROVED** [review]
Hey @shoom3301 , great, thank you!
Just a nitpick: I noticed that the banner appears in 1-2 minutes, so a user is able to place an order. See the video: https://www.loom.com/share/ce20825a5abc4e21af1d025a867cb5ce

LMK please if it worth addressing in a separate PR.

File: apps/cowswap-frontend/src/common/containers/InvalidLocalTimeWarning/index.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  const localTimeOffset = useLocalTimeOffset()
+
+  if (!localTimeOffset || localTimeOffset < TIME_OFFSET_THRESHOLD) return null
+
+  console.debug('Local time offset:', localTimeOffset)
+
+  return (
+    <GlobalWarning>
+      Local device time does not match real time, CoW Swap most likely will not work correctly. Please synchronize the
+      time settings on your device with the real time.
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/common/containers/InvalidLocalTimeWarning/index.tsx:20]: Sentence could be shorter.
Maybe:

```suggestion
      Local device time does is not accurate, CoW Swap most likely will not work correctly. Please adjust your device's time.
```

Not very happy with that either.

@MindyCoW ideas?

**Review by alfetopito — APPROVED** [review]

**Pull request merged**